RedHatInsights/compliance-frontend
# refactor: Refactor/clean AsyncTable hooks and add documentation

This PR extends the documentation for AsyncTable hooks as well as how to use them and some commonly used components/functions/hooks in compliance.

It also adds a linter for React hooks rules and warns about common errors. The hooks have been updated to comply to these rules (as best as possible). 

For "complex" objects that are used as dependencies for React hooks the `useDeepCompareEffect(NoCheck)` hook is used to avoid unnecessary re-renders. In cases with complex objects where this hook can not be use `JSON.stringify` is used (for now).

The `useTableState` has been extended to support "observers" to allow a more descriptive way of defining interactions, like the pagination resetting to page 1 filters or sorting changes. (See tests for example).

The `useItems` hook has been extracted into its folder, with separate tests, as it is becoming more important. It has also been updated to use a table state to make items in the table accessible via the context and allow interaction with other hooks. It also now has a `loaded` state, that is not really utilised for now, but is accessible via the context and set properly when items are "loaded".

**How to test:**

1) Run `npm run docs`
3) Verify the documentation generated successfully.
2) Open `docs/index.html`
3) (Proof)read the functions documentation on options, and returns and the tutorial(s).
4) Start the dev proxy and enable AsyncTable
5) Open the Reports page
6) Verify that table is properly rendered and interactions are working

## Secure Coding Practices Checklist GitHub Link
- https://github.com/RedHatInsights/secure-coding-checklist

## Secure Coding Checklist
